File: backend/app/routers/answers.py
import asyncio
import json
import logging
import os
import uuid as uuid_module
from uuid import UUID
from fastapi import APIRouter, Depends, File, Form, HTTPException, UploadFile, status
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession
from app.config import get_settings
from app.db.database import get_db
from app.dependencies import get_current_user
from app.models.user import User
from app.models.session import Session
from app.models.exchange import Exchange
from app.models.integrity import Score
from app.schemas.schemas import (
    AnswerSubmit, AnswerProcessingResponse, EvaluationResult,
    AudioAnswerProcessingResponse, SpeechAnalysisResponse,
)
from app.services.evaluator_agent import evaluate_answer
from app.services.interviewer_agent import generate_next_question
from app.services.transcription_service import transcribe_audio
from app.services.vector_memory import (
    embed_and_store_exchange,
    retrieve_relevant_weak_answers,
    format_retrieved_context,
)
from app.services.weak_topic_tracker import update_weak_topics
from app.services.speech_analyzer import analyze_speech

logger = logging.getLogger(__name__)
settings = get_settings()

router = APIRouter(prefix="/answers", tags=["Answers"])


# ─── Helper: common answer processing logic ──────────────────

async def _process_evaluation_and_next_question(
    session: Session,
    current_exchange: Exchange,
    transcript: str,
    db: AsyncSession,
) -> tuple[EvaluationResult, str | None, bool]:
    """
    Run evaluator on transcript, store score, generate next question.
    Returns (evaluation, next_question_text, session_complete).
    """
    # Run evaluator agent
    evaluation = await evaluate_answer(
        question=current_exchange.question,
        transcript=transcript,
        topic=session.topic,
    )

    # Store the score (speech metrics will be added by the caller if available)
    score = Score(
        exchange_id=current_exchange.id,
        technical_accuracy=evaluation.technical_accuracy,
        definition_present=evaluation.definition_present,
        mechanism_explained=evaluation.mechanism_explained,
        example_given=evaluation.example_given,
        edge_cases_mentioned=evaluation.edge_cases_mentioned,
        missing_concepts=json.dumps(evaluation.missing_concepts) if evaluation.missing_concepts else None,
        follow_up_angle=evaluation.follow_up_angle,
    )
    db.add(score)
    await db.flush()

    # ─── Week 3: Embed exchange + update weak topics (non-blocking) ───
    try:
        await embed_and_store_exchange(
            exchange_id=current_exchange.id,
            question=current_exchange.question,
            answer=transcript,
            db=db,
        )
    except Exception as e:
        logger.warning(f"Embedding storage failed (non-critical): {e}")

    try:
        await update_weak_topics(
            user_id=session.user_id,
            topic=session.topic,
            evaluation=evaluation,
            db=db,
        )
    except Exception as e:
        logger.warning(f"Weak topic update failed (non-critical): {e}")

    # Check if session is complete
    next_index = current_exchange.question_index + 1
    session_complete = next_index > session.total_questions

    next_question_text = None

    if not session_complete:
        # Build performance summary
        all_exchanges_result = await db.execute(
            select(Exchange)
            .where(Exchange.session_id == session.id)
            .order_by(Exchange.question_index)
        )
        all_exchanges = all_exchanges_result.scalars().all()

        scores_list = []
        for ex in all_exchanges:
            if ex.score and ex.score.technical_accuracy is not None:
                scores_list.append(ex.score.technical_accuracy)

        avg_score = sum(scores_list) / len(scores_list) if scores_list else 0
        performance_summary = (
            f"Questions answered: {len(scores_list)}/{session.total_questions}. "
            f"Average score: {avg_score:.1f}/10."
        )

        # ─── Week 3: Retrieve cross-session memory context ────────
        retrieved_context = ""
        try:
            weak_answers = await retrieve_relevant_weak_answers(
                user_id=session.user_id,
                current_question=current_exchange.question,
                db=db,
            )
            retrieved_context = format_retrieved_context(weak_answers)
            if retrieved_context:
                logger.info(f"Injecting {len(weak_answers)} retrieved weak answers into interviewer context")
        except Exception as e:
            logger.warning(f"Memory retrieval failed (non-critical): {e}")

        # Generate next question (with memory context)
        next_question_text = await generate_next_question(
            topic=session.topic,
            difficulty=session.difficulty,
            evaluation=evaluation,
            question_index=next_index,
            total_questions=session.total_questions,
            performance_summary=performance_summary,
            retrieved_context=retrieved_context,
        )

        # Create next exchange
        next_exchange = Exchange(
            session_id=session.id,
            question=next_question_text,
            question_index=next_index,
        )
        db.add(next_exchange)
        await db.flush()
    else:
        # Mark session as completed
        from datetime import datetime, timezone
        session.status = "completed"
        session.ended_at = datetime.now(timezone.utc)
        await db.flush()

    return evaluation, next_question_text, session_complete


async def _get_active_exchange(session: Session, db: AsyncSession) -> Exchange:
    """Get the current unanswered exchange for a session."""
    result = await db.execute(
        select(Exchange)
        .where(
            Exchange.session_id == session.id,
            Exchange.answer_transcript.is_(None),
        )
        .order_by(Exchange.question_index)
        .limit(1)
    )
    exchange = result.scalar_one_or_none()

    if not exchange:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="No pending question found",
        )

    return exchange


async def _validate_session(session_id: UUID, user: User, db: AsyncSession) -> Session:
    """Verify session belongs to user and is active."""
    result = await db.execute(
        select(Session).where(
            Session.id == session_id,
            Session.user_id == user.id,
        )
    )
    session = result.scalar_one_or_none()

    if not session:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Session not found",
        )

    if session.status != "active":
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Session is not active",
        )

    return session


# ─── Endpoint: Text Answer (Week 1 — preserved) ──────────────

@router.post("/submit", response_model=AnswerProcessingResponse)
async def submit_answer(
    request: AnswerSubmit,
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    """
    Submit a text answer for the current question.
    Runs evaluator agent → stores score → generates next question.
    """
    session = await _validate_session(request.session_id, user, db)
    current_exchange = await _get_active_exchange(session, db)

    # Store the answer transcript
    current_exchange.answer_transcript = request.answer_text
    await db.flush()

    # Process evaluation and next question
    evaluation, next_question_text, session_complete = (
        await _process_evaluation_and_next_question(
            session, current_exchange, request.answer_text, db
        )
    )

    return AnswerProcessingResponse(
        exchange_id=current_exchange.id,
        evaluation=evaluation,
        next_question=next_question_text,
        question_index=current_exchange.question_index,
        session_complete=session_complete,
    )


# ─── Endpoint: Audio Answer (Week 2 — Voice Pipeline) ────────

@router.post("/submit-audio", response_model=AudioAnswerProcessingResponse)
async def submit_audio_answer(
    audio: UploadFile = File(..., description="Audio recording of the answer"),
    session_id: str = Form(..., description="Session UUID"),
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    """
    Submit an audio answer for the current question.
    Runs parallel pipeline:
      1. Whisper API → transcript + word timestamps
      2. Speech analysis → WPM, pauses, fillers
      3. (Placeholder) Speaker verification
    Then sequential:
      4. Evaluator agent on transcript → structured score
      5. Generate next question
    """
    # Parse and validate session_id
    try:
        parsed_session_id = UUID(session_id)
    except ValueError:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Invalid session_id format",
        )

    session = await _validate_session(parsed_session_id, user, db)
    current_exchange = await _get_active_exchange(session, db)

    # Validate audio file
    if not audio.filename:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="No audio file provided",
        )

    # Check file size
    content = await audio.read()
    max_size = settings.MAX_AUDIO_SIZE_MB * 1024 * 1024
    if len(content) > max_size:
        raise HTTPException(
            status_code=status.HTTP_413_REQUEST_ENTITY_TOO_LARGE,
            detail=f"Audio file too large. Maximum: {settings.MAX_AUDIO_SIZE_MB}MB",
        )

    # Save audio to local filesystem
    audio_dir = os.path.join(settings.UPLOAD_DIR, "audio", str(parsed_session_id))
    os.makedirs(audio_dir, exist_ok=True)

    # Determine file extension from content type or filename
    ext = ".webm"
    if audio.content_type:
        content_type_map = {
            "audio/webm": ".webm",
            "audio/ogg": ".ogg",
            "audio/wav": ".wav",
            "audio/mp4": ".mp4",
            "audio/mpeg": ".mp3",
        }
        ext = content_type_map.get(audio.content_type, ".webm")
    elif audio.filename and "." in audio.filename:
        ext = "." + audio.filename.rsplit(".", 1)[1]

    audio_filename = f"{current_exchange.id}{ext}"
    audio_path = os.path.join(audio_dir, audio_filename)

    with open(audio_path, "wb") as f:
        f.write(content)

    logger.info(
        f"Audio saved: {audio_path} ({len(content)} bytes, {audio.content_type})"
    )

    # ─── Parallel Processing Pipeline ─────────────────────────
    # Run transcription and (future) speaker verification in parallel.
    # Speech analysis depends on transcription output, so it runs after.

    transcription_result = await transcribe_audio(audio_path)

    # Run speech analysis on the transcription result
    speech_analysis = analyze_speech(
        word_timestamps=transcription_result.word_timestamps,
        transcript=transcription_result.transcript,
    )

    # Store the answer transcript
    current_exchange.answer_transcript = transcription_result.transcript
    await db.flush()

    # ─── Sequential: Evaluation + Next Question ───────────────

    evaluation, next_question_text, session_complete = (
        await _process_evaluation_and_next_question(
            session, current_exchange, transcription_result.transcript, db
        )
    )

    # Update the score with speech metrics
    score_result = await db.execute(
        select(Score).where(Score.exchange_id == current_exchange.id)
    )
    score = score_result.scalar_one_or_none()
    if score:
        score.wpm = int(speech_analysis.avg_wpm)
        score.filler_count = speech_analysis.filler_count
        score.longest_pause_seconds = speech_analysis.longest_pause_seconds
        score.confidence_proxy = speech_analysis.confidence_proxy
        await db.flush()

    # Build speech analysis response
    speech_response = SpeechAnalysisResponse(
        avg_wpm=speech_analysis.avg_wpm,
        wpm_std_dev=speech_analysis.wpm_std_dev,
        total_duration=speech_analysis.total_duration,
        word_count=speech_analysis.word_count,
        pause_count=speech_analysis.pause_count,
        longest_pause_seconds=speech_analysis.longest_pause_seconds,
        filler_count=speech_analysis.filler_count,
        filler_words=speech_analysis.filler_words,
        confidence_proxy=speech_analysis.confidence_proxy,
    )

    return AudioAnswerProcessingResponse(
        exchange_id=current_exchange.id,
        transcript=transcription_result.transcript,
        evaluation=evaluation,
        speech_analysis=speech_response,
        next_question=next_question_text,
        question_index=current_exchange.question_index,
        session_complete=session_complete,
    )

File: backend/src/routes/answers.ts
import { Router, Response } from 'express';
import { v4 as uuidv4 } from 'uuid';
import path from 'path';
import fs from 'fs';
import multer from 'multer';
import { getDb } from '../db';
import { requireAuth, AuthRequest } from '../middleware/auth';
import { evaluateAnswer, EvaluationResult } from '../services/evaluatorAgent';
import { generateNextQuestion } from '../services/interviewerAgent';
import { transcribeAudio } from '../services/transcriptionService';
import { analyzeSpeech } from '../services/speechAnalyzer';
import { embedAndStoreExchange, retrieveRelevantWeakAnswers, formatRetrievedContext } from '../services/vectorMemory';
import { updateWeakTopics } from '../services/weakTopicTracker';
import { computeGazeFluencyCorrelation, updateIntegrityScore } from '../services/integrityEngine';
import { config } from '../config';

const router = Router();

// Setup multer for temporary audio uploads
const upload = multer({
  dest: path.resolve(config.UPLOAD_DIR, 'temp'),
  limits: { fileSize: config.MAX_AUDIO_SIZE_MB * 1024 * 1024 }
});

// Helper: Common answer evaluation and next question logic
async function processEvaluationAndNextQuestion(
  session: any,
  currentExchange: any,
  transcript: string,
  user: any
): Promise<{ evaluation: EvaluationResult; nextQuestionText: string | null; sessionComplete: boolean }> {
  const db = getDb();

  const provider = session.llm_provider || 'anthropic';
  const model = session.llm_model || 'claude-sonnet-4-20250514';

  // 1. Run technical evaluation
  const evaluation = await evaluateAnswer(currentExchange.question, transcript, session.topic, provider, model, user);

  // 2. Insert core Score record
  const scoreId = uuidv4();
  await db.run(
    `INSERT INTO scores 
      (id, exchange_id, technical_accuracy, definition_present, mechanism_explained, example_given, edge_cases_mentioned, missing_concepts, follow_up_angle) 
     VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?)`,
    [
      scoreId,
      currentExchange.id,
      evaluation.technical_accuracy,
      evaluation.definition_present ? 1 : 0,
      evaluation.mechanism_explained ? 1 : 0,
      evaluation.example_given ? 1 : 0,
      evaluation.edge_cases_mentioned ? 1 : 0,
      evaluation.missing_concepts ? JSON.stringify(evaluation.missing_concepts) : null,
      evaluation.follow_up_angle
    ]
  );

  // 3. Vector memory storage and weak topic tracking in "background" (non-blocking)
  embedAndStoreExchange(currentExchange.id, currentExchange.question, transcript, user)
    .catch(err => console.error('Background embedding storage failed:', err));

  updateWeakTopics(session.user_id, session.topic, evaluation)
    .catch(err => console.error('Background weak topic tracker failed:', err));

  // 4. Determine next step
  const nextIndex = currentExchange.question_index + 1;
  const sessionComplete = nextIndex > session.total_questions;

  let nextQuestionText: string | null = null;

  if (!sessionComplete) {
    // Build running performance summary
    const allExchanges = await db.all(
      `SELECT e.id, s.technical_accuracy 
       FROM exchanges e
       LEFT JOIN scores s ON s.exchange_id = e.id
       WHERE e.session_id = ?
       ORDER BY e.question_index ASC`,
      [session.id]
    );

    const scoresList = allExchanges
      .map(ex => ex.technical_accuracy)
      .filter((s): s is number => s !== null && s !== undefined);

    const avgScore = scoresList.length > 0 ? scoresList.reduce((a, b) => a + b, 0) / scoresList.length : 0.0;
    const performanceSummary =
      `Questions answered: ${scoresList.length}/${session.total_questions}. ` +
      `Average score: ${avgScore.toFixed(1)}/10.`;

    // Retrieve weak answers from memory
    let retrievedContext = '';
    try {
      const weakAnswers = await retrieveRelevantWeakAnswers(session.user_id, currentExchange.question, user);
      retrievedContext = formatRetrievedContext(weakAnswers);
    } catch (err) {
      console.warn('Memory retrieval failed (non-critical):', err);
    }

    // Generate next question
    nextQuestionText = await generateNextQuestion(
      session.topic,
      session.difficulty,
      evaluation,
      nextIndex,
      session.total_questions,
      performanceSummary,
      retrievedContext,
      provider,
      model,
      user
    );

    // Create next exchange
    const nextExchangeId = uuidv4();
    const now = new Date().toISOString();
    await db.run(
      'INSERT INTO exchanges (id, session_id, question, question_index, created_at) VALUES (?, ?, ?, ?, ?)',
      [nextExchangeId, session.id, nextQuestionText, nextIndex, now]
    );
  } else {
    // End the session
    const endedAt = new Date().toISOString();
    await db.run('UPDATE sessions SET status = ?, ended_at = ? WHERE id = ?', ['completed', endedAt, session.id]);
  }

  return { evaluation, nextQuestionText, sessionComplete };
}

// GET active unanswered exchange
async function getActiveExchange(sessionId: string): Promise<any> {
  const db = getDb();
  const exchange = await db.get(
    'SELECT * FROM exchanges WHERE session_id = ? AND answer_transcript IS NULL ORDER BY question_index ASC LIMIT 1',
    [sessionId]
  );
  return exchange;
}

// Validate session active and owned by user
async function validateSession(sessionId: string, userId: string): Promise<any> {
  const db = getDb();
  const session = await db.get('SELECT * FROM sessions WHERE id = ? AND user_id = ?', [sessionId, userId]);
  if (!session) {
    const err: any = new Error('Session not found');
    err.status = 404;
    throw err;
  }
  if (session.status !== 'active') {
    const err: any = new Error('Session is not active');
    err.status = 400;
    throw err;
  }
  return session;
}

// POST /answers/submit
router.post('/submit', requireAuth, async (req: AuthRequest, res: Response) => {
  const { session_id, answer_text } = req.body;

  if (!session_id || !answer_text) {
    return res.status(400).json({ detail: 'session_id and answer_text are required' });
  }

  try {
    const session = await validateSession(session_id, req.userId!);
    const currentExchange = await getActiveExchange(session_id);

    if (!currentExchange) {
      return res.status(400).json({ detail: 'No pending question found' });
    }

    const db = getDb();

    // Store transcript
    await db.run('UPDATE exchanges SET answer_transcript = ? WHERE id = ?', [answer_text, currentExchange.id]);

    const { evaluation, nextQuestionText, sessionComplete } = await processEvaluationAndNextQuestion(
      session,
      currentExchange,
      answer_text,
      req.user
    );

    return res.json({
      exchange_id: currentExchange.id,
      evaluation,
      next_question: nextQuestionText,
      question_index: currentExchange.question_index,
      session_complete: sessionComplete,
    });
  } catch (error: any) {
    console.error('Text submission error:', error);
    const status = error.status || 500;
    return res.status(status).json({ detail: error.message || 'Internal server error submitting text answer' });
  }
});

// POST /answers/submit-audio
router.post('/submit-audio', requireAuth, upload.single('audio'), async (req: AuthRequest, res: Response) => {
  const { session_id } = req.body;

  if (!session_id) {
    if (req.file) fs.unlinkSync(req.file.path);
    return res.status(400).json({ detail: 'session_id is required' });
  }

  if (!req.file) {
    return res.status(400).json({ detail: 'No audio file provided' });
  }

  try {
    const session = await validateSession(session_id, req.userId!);
    const currentExchange = await getActiveExchange(session_id);

    if (!currentExchange) {
      fs.unlinkSync(req.file.path);
      return res.status(400).json({ detail: 'No pending question found' });
    }

    // Determine path extension
    let ext = '.webm';
    if (req.file.mimetype) {
      const mimeMap: Record<string, string> = {
        'audio/webm': '.webm',
        'audio/ogg': '.ogg',
        'audio/wav': '.wav',
        'audio/mp4': '.mp4',
        'audio/mpeg': '.mp3',
      };
      ext = mimeMap[req.file.mimetype] || '.webm';
    }

    const audioDir = path.join(config.UPLOAD_DIR, 'audio', session_id);
    fs.mkdirSync(audioDir, { recursive: true });

    const audioPath = path.join(audioDir, `${currentExchange.id}${ext}`);
    fs.renameSync(req.file.path, audioPath);

    console.log(`Audio saved: ${audioPath} (${req.file.size} bytes)`);

    // ─── Parallel Processing Pipeline ───
    const transcriptionResult = await transcribeAudio(audioPath, req.user);

    const speechAnalysis = analyzeSpeech(
      transcriptionResult.word_timestamps,
      transcriptionResult.transcript
    );

    const db = getDb();
    await db.run('UPDATE exchanges SET answer_transcript = ? WHERE id = ?', [
      transcriptionResult.transcript,
      currentExchange.id,
    ]);

    // Calculate gaze-fluency correlation and update score
    let correlation = 0.0;
    try {
      correlation = await computeGazeFluencyCorrelation(
        session.id,
        currentExchange.id,
        speechAnalysis.wpm_segments
      );
    } catch (err) {
      console.error('Error computing correlation:', err);
    }

    try {
      await updateIntegrityScore(session.id, correlation, currentExchange.id);
    } catch (err) {
      console.error('Error updating integrity score:', err);
    }

    // ─── Sequential: Evaluation + Next Question ───
    const { evaluation, nextQuestionText, sessionComplete } = await processEvaluationAndNextQuestion(
      session,
      currentExchange,
      transcriptionResult.transcript,
      req.user
    );

    // Save speech stats into score record
    await db.run(
      `UPDATE scores 
       SET wpm = ?, filler_count = ?, longest_pause_seconds = ?, confidence_proxy = ?, gaze_fluency_correlation = ? 
       WHERE exchange_id = ?`,
      [
        Math.floor(speechAnalysis.avg_wpm),
        speechAnalysis.filler_count,
        speechAnalysis.longest_pause_seconds,
        speechAnalysis.confidence_proxy,
        correlation,
        currentExchange.id,
      ]
    );

    // Format speech analysis response properties matching Python models
    const speechResponse = {
      avg_wpm: speechAnalysis.avg_wpm,
      wpm_std_dev: speechAnalysis.wpm_std_dev,
      total_duration: speechAnalysis.total_duration,
      word_count: speechAnalysis.word_count,
      pause_count: speechAnalysis.pause_count,
      longest_pause_seconds: speechAnalysis.longest_pause_seconds,
      filler_count: speechAnalysis.filler_count,
      filler_words: speechAnalysis.filler_words,
      confidence_proxy: speechAnalysis.confidence_proxy,
    };

    return res.json({
      exchange_id: currentExchange.id,
      transcript: transcriptionResult.transcript,
      evaluation,
      speech_analysis: speechResponse,
      next_question: nextQuestionText,
      question_index: currentExchange.question_index,
      session_complete: sessionComplete,
    });
  } catch (error: any) {
    console.error('Audio submission error:', error);
    if (req.file && fs.existsSync(req.file.path)) {
      fs.unlinkSync(req.file.path);
    }
    const status = error.status || 500;
    return res.status(status).json({ detail: error.message || 'Internal server error submitting audio answer' });
  }
});

export default router;
